### DISCUSSION TITLE
[Help] Getting embeddings of ara-gpt2

### DISCUSSION THREAD

**Discussion by kwcckw** [aub-mind/arabert#112] [Q&A] [answered]
Hi,

Good day.

i'm trying to get the embedding output of ara-gpt2 but looks like facing error：

![image](https://user-images.githubusercontent.com/52521979/121328039-91721a80-c946-11eb-9e3c-2cf66d60857a.png)

And here's the code:

        import torch
        
        from transformers import GPT2TokenizerFast, pipeline
        #for base and medium
        from transformers import GPT2LMHeadModel
        #for large and mega
        from aragpt2.grover.modeling_gpt2 import GPT2LMHeadModel
        
        from preprocess import ArabertPreprocessor
        
        device = 0 if torch.cuda.is_available() else -1
        
        MODEL_NAME='aragpt2-base'
        arabert_prep = ArabertPreprocessor(model_name=MODEL_NAME)
        
        text1  = 'ذهب نوح مظفر ضرغام بصحبة رؤوف بن لؤي ظافر عطعوط وهلال  ';
        
        text_clean = arabert_prep.preprocess(text1)
        
        model = GPT2LMHeadModel.from_pretrained(MODEL_NAME)
        tokenizer = GPT2TokenizerFast.from_pretrained(MODEL_NAME)
        generation_pipeline = pipeline("text-generation",model=model,tokenizer=tokenizer, device=device)
        
        input_size = len(tokenizer.tokenize(text_clean))
        
        gen_text = generation_pipeline(text_clean,
                    pad_token_id=0, # 0 for AraGPT2
                    do_sample = False,
                    num_beams=5,
                    max_length=input_size + 5,
                    top_k=10,
                    top_p=0.95,
                    repetition_penalty = 3.0,
                    no_repeat_ngram_size = 3,
                    num_return_sequences = 5)
        
        inputs = tokenizer.encode_plus(text_clean, return_tensors='pt')
        
        print(tokenizer.convert_ids_to_tokens(inputs['input_ids'][0]))
        
        outputs = model(**inputs.to(device))
        
        embeddings = outputs['last_hidden_state']
        
        
i checked and doesn't see any 'last_hidden_state' in the outputs too:

![image](https://user-images.githubusercontent.com/52521979/121328161-acdd2580-c946-11eb-85f4-bef22ce7f1b1.png)

Could you advise on how to get the embedding?

Thank you.

**Comment by WissamAntoun** [comment, marked as answer]
check the following notebook https://colab.research.google.com/drive/17B3bY38CW5VeWkfJsLOaxq6pzHtHKc5W?usp=sharing for how to extract embeddings from GPT2 models. Also please be careful from where are you import the GPT2Model class (use the huggingface library for aragpt2-base and medium,  and from arabert.grover... for large and mega)

**Reply by kwcckw** [reply]
Thanks, looks like i need to enable the hidden state to do allow it:

    model = GPT2LMHeadModel.from_pretrained(MODEL_NAME,output_hidden_states=True)
